### PR TITLE
Group refactoring

### PR DESCRIPTION
**Pull request by swilly22**
Switching from `rax` to `hashtable` for aggregation groups lookup

### CODE REVIEW

**Comment by filipecosta90** [comment]
### Automated performance analysis summary

This comment was automatically generated given there is performance data available.

In summary:
- Detected a total of 17 stable tests between versions.
- Detected a total of 1 highly unstable benchmarks.
- Detected a total of 1 regressions bellow the regression water line 5.0.

You can check a comparison in detail via the [grafana link](https://benchmarksrediscom.grafana.net/d//SH9_rQYGz/?var-branch=master&var-branch=group-refactoring)

### Comparison between master and group-refactoring.

Time Period from 30 days ago. (environment used: oss-standalone)

|                               Test Case                               |Baseline master (median obs. +- std.dev)|Comparison group-refactoring (median obs. +- std.dev)|% change (higher-better)|             Note              |
|-----------------------------------------------------------------------|----------------------------------------|-----------------------------------------------------|------------------------|-------------------------------|
|ENTITY_COUNT                                                           | 33324 +- 0.0%  (7 datapoints)          | 33323 +- nan%  (1 datapoints)                       |-0.0%                   |-- no change --                |
|EXPLICIT-EDGE-DELETION                                                 | 100 +- 0.0%  (7 datapoints)            | 100 +- nan%  (1 datapoints)                         |-0.0%                   |-- no change --                |
|GRAPH500-SCALE_18-EF_16-1_HOP                                          | 19999 +- 6.3%  (7 datapoints)          | 20000 +- nan%  (1 datapoints)                       |0.0%                    |waterline=6.3%. -- no change --|
|GRAPH500-SCALE_18-EF_16-1_HOP_MIXED_READ_65perc_WRITE_25perc_DEL_10perc| 50 +- 2.0%  (7 datapoints)             | 50 +- nan%  (1 datapoints)                          |-0.0%                   |-- no change --                |
|GRAPH500-SCALE_18-EF_16-1_HOP_MIXED_READ_75perc_WRITE_25perc           | 59 +- 0.0%  (7 datapoints)             | 59 +- nan%  (1 datapoints)                          |-0.0%                   |-- no change --                |
|GRAPH500-SCALE_18-EF_16-2_HOP                                          | 16665 +- 0.0%  (7 datapoints)          | 16665 +- nan%  (1 datapoints)                       |-0.0%                   |-- no change --                |
|GRAPH500-SCALE_18-EF_16-2_HOP_MIXED_READ_75perc_WRITE_25perc           | 59 +- 4.0%  (7 datapoints)             | 59 +- nan%  (1 datapoints)                          |-0.0%                   |-- no change --                |
|GRAPH500-SCALE_18-EF_16-3_HOP                                          | 4348 +- 0.0%  (7 datapoints)           | 3846 +- nan%  (1 datapoints)                        |-11.5%                  |REGRESSION                     |
|GRAPH500-SCALE_18-EF_16-3_HOP_MIXED_READ_75perc_WRITE_25perc           | 56 +- 2.0%  (7 datapoints)             | 56 +- nan%  (1 datapoints)                          |0.0%                    |-- no change --                |
|IMPLICIT-EDGE-DELETION                                                 | 17 +- 0.0%  (7 datapoints)             | 17 +- nan%  (1 datapoints)                          |0.0%                    |-- no change --                |
|INSPECT-EDGES                                                          | 109 +- 11.7% UNSTABLE (7 datapoints)   | 116 +- nan%  (1 datapoints)                         |7.0%                    |UNSTABLE (very high variance)  |
|MIX_READ_WRITE                                                         | 11111 +- 3.8%  (7 datapoints)          | 11111 +- nan%  (1 datapoints)                       |-0.0%                   |-- no change --                |
|NODE-INDEX-LOOKUP                                                      | 33 +- 0.0%  (7 datapoints)             | 33 +- nan%  (1 datapoints)                          |0.0%                    |-- no change --                |
|SORT_ENTITIES                                                          | 100 +- 0.0%  (7 datapoints)            | 100 +- nan%  (1 datapoints)                         |0.0%                    |-- no change --                |
|UPDATE-BASELINE                                                        | 24999 +- 0.0%  (7 datapoints)          | 24998 +- nan%  (1 datapoints)                       |-0.0%                   |-- no change --                |
|VARIABLE-LENGTH-FILTER                                                 | 33326 +- 0.0%  (7 datapoints)          | 33325 +- nan%  (1 datapoints)                       |-0.0%                   |-- no change --                |
|VARIABLE_LENGTH_EXPAND_INTO                                            | 100 +- 0.0%  (7 datapoints)            | 100 +- nan%  (1 datapoints)                         |-0.0%                   |-- no change --                |
|allShortestPaths-10hop-100Kpaths                                       | 16 +- 2.1%  (7 datapoints)             | 16 +- nan%  (1 datapoints)                          |-0.0%                   |-- no change --                |
|allShortestPaths-4hop-10Kpaths                                         | 208 +- 2.9%  (7 datapoints)            | 208 +- nan%  (1 datapoints)                         |0.0%                    |-- no change --                |

File: src/execution_plan/ops/op_aggregate.c [hunk 1]
```diff
@@ -159,72 +180,84 @@ static void _aggregateRecord(OpAggregate *op, Record r) {
 		AR_ExpNode *exp = group->aggregationFunctions[i];
 		AR_EXP_Aggregate(exp, r);
 	}
-
-	// free record
-	OpBase_DeleteRecord(r);
```
Comment (AviAvni) [inline src/execution_plan/ops/op_aggregate.c:164]: why remove this? where we delete the record?

File: src/execution_plan/execution_plan_build/build_projection_ops.c [hunk 2]
```diff
@@ -112,6 +113,12 @@ static void _combine_projection_arrays(AR_ExpNode ***exps_ptr, AR_ExpNode **orde
 		raxTryInsert(projection_names, (unsigned char *)name, strlen(name), NULL, NULL);
 	}
 
+	// check if the projection expressions contain aggregation an expression
+	bool is_agg = false;
+	for(uint i = 0; i < project_count && !is_agg; i++) {
+		is_agg = AR_EXP_ContainsAggregation(project_exps[i]);
+	}
+
```
Comment (swilly22) [inline src/execution_plan/execution_plan_build/build_projection_ops.c:121]: Why do we need this second loop,
simply perform this assignment on the previous loop,
also update the assignment to: `is_agg |= AR_EXP_ContainsAggregation(project_exps[i]);`

File: tests/flow/test_with_clause.py [hunk 3]
```diff
@@ -267,3 +267,13 @@ def test12_cartesian_product_reset_single_response(self):
...
         self.env.assertEqual(len(actual_result.result_set), 36)
+
+    def test13_aggregated_projection_vars(self):
+        query = "WITH 1 as one UNWIND range(0,4) as x RETURN max(x) ORDER BY one"
+        try:
+            redis_graph.query(query)
+        except redis.exceptions.ResponseError as e:
+            # error expected
+            self.env.assertIn(
+                "only projections are valid in aggregated projections ordering",
```
Comment (swilly22) [inline tests/flow/test_with_clause.py:278]: Please see Neo's error message and adopt a similar version

**Review by AviAvni — DISMISSED** [review]

**Pull request merged**